tensorboard/plugins/histogram/histograms_plugin.py

```diff
@@ -137,7 +162,30 @@ def histograms_impl(self, tag, run, downsample_to=None):
     Raises:
       tensorboard.errors.PublicError: On invalid request.
     """
-    if self._db_connection_provider:
+    if self._data_provider:
+      # Downsample reads to 500 histograms per time series, which is
+      # the default size guidance for histograms under the multiplexer
+      # loading logic.
+      SAMPLE_COUNT = downsample_to if downsample_to is not None else 500
+      all_histograms = self._data_provider.read_tensors(
+          experiment_id=experiment,
+          plugin_name=metadata.PLUGIN_NAME,
+          downsample=SAMPLE_COUNT,
+          run_tag_filter=provider.RunTagFilter(runs=[run], tags=[tag]),
+      )
+      histograms = all_histograms.get(run, {}).get(tag, None)
+      if histograms is None:
+        raise errors.NotFoundError(
+            "No histogram tag %r for run %r" % (tag, run)
+        )
+      if downsample_to is not None:
+        rng = random.Random(0)
+        histograms = _downsample(rng, histograms, downsample_to)
```

Comment:
Yeah, I did this for consistency with the existing code so that the
responses are exactly identical before and after this change. The data
provider is *supposed* to downsample, but actually doesn’t, so this is
necessary. Without this, we see 500 histograms instead of 51 in the UI.

You’re right that this is a bit confusing; I’ll add a comment. Thanks.

Reply:
Done.